llvm/lib/Analysis/BasicAliasAnalysis.cpp

```diff
@@ -663,22 +664,56 @@ BasicAAResult::DecomposeGEPExpression(const Value *V, const DataLayout &DL,
       unsigned Width = Index->getType()->getIntegerBitWidth();
       unsigned SExtBits = IndexSize > Width ? IndexSize - Width : 0;
       unsigned TruncBits = IndexSize < Width ? Width - IndexSize : 0;
-      LinearExpression LE = GetLinearExpression(
-          CastedValue(Index, 0, SExtBits, TruncBits), DL, 0, AC, DT);
+      // Scalable GEP decomposition
+      // Allow Scalable GEP to be decomposed in the case of
+      //    1. getelementptr <4 x vscale x i32> with 1st index as a constant
+      //    2. Index which have a leaf of @llvm.vscale
+      // In both cases, essentially CastedValue of VariableGEPIndex is Vscale,
+      // however in the 1st case, CastedValue is of type constant, hence another
+      // flag in VariableGEPIndex is created in this case, IsVScale If GEP is
+      // Scalable type, e.g. <4 x vscale x i32>, the first index will have
+      // vscale as a variable index, create a LE in this case.
```

Comment:
So it sounds like you are interested in modelling `vscale * scale` offsets, but not `vscale * scale * var`. If IsVScale is set, then the variable is just ignored, is that correct?

If so, then an alternative way to model this is to use some kind of sentinel value (say `reinterpret_cast<Value*>(-1)`) as the variable, which will get interpreted as vscale. Then, we should be able to reuse the existing code, with some adjustments to e.g. make the range calculation for the value use the vscale range, etc.

Reply:
Okay, looking a bit further I guess you kind of already do this, just with checks of the IsVScale flag. I think I'd still prefer a sentinel value, to make sure we "crash and burn" if we fail to handle it somewhere, instead of silently miscompiling. This should also allow you to do something actually useful for the vscale intrinsinc handling in LinearExpression, as we would ensure that the same value gets used for `vscale x` and the intrinsic.

Reply:
I think that sounds like a good idea, but it is currently using the value for the bitwidth so may need some other way to represent that.